File: python/src/revenant/ui/cli/verify.py
# pyright: reportUnknownMemberType=false, reportUnknownVariableType=false, reportUnknownArgumentType=false
"""
Signature verification and inspection.

cmd_verify uses openssl CLI (graceful degradation if not installed).
cmd_info uses asn1crypto (cross-platform, no external tools).
"""

from __future__ import annotations

import subprocess
import sys
from pathlib import Path
from typing import TYPE_CHECKING

from asn1crypto import cms as asn1_cms

if TYPE_CHECKING:
    import argparse


def cmd_verify(args: argparse.Namespace) -> None:
    """Verify a detached CMS signature using openssl."""
    pdf_path = Path(args.pdf)
    sig_path = Path(args.signature) if args.signature else pdf_path.with_suffix(".pdf.p7s")

    if not pdf_path.exists():
        print(f"Error: {pdf_path} not found", file=sys.stderr)
        sys.exit(1)
    if not sig_path.exists():
        print(f"Error: {sig_path} not found", file=sys.stderr)
        sys.exit(1)

    print(f"Verifying {pdf_path.name} against {sig_path.name}...")

    cmd = [
        "openssl",
        "cms",
        "-verify",
        "-inform",
        "DER",
        "-in",
        str(sig_path),
        "-content",
        str(pdf_path),
        "-binary",
        "-purpose",
        "any",
    ]

    print("  Using system trust store for chain verification")

    try:
        result = subprocess.run(cmd, capture_output=True, timeout=15)
        stderr = result.stderr.decode("utf-8", errors="replace").strip()

        if result.returncode == 0:
            print("  VALID: Signature verification succeeded.")
            if "Verification successful" in stderr:
                print(f"  {stderr}")
        else:
            print(f"  INVALID: {stderr}")
            sys.exit(1)
    except FileNotFoundError:
        print("Error: openssl not found. Install OpenSSL to verify signatures.", file=sys.stderr)
        sys.exit(1)
    except subprocess.TimeoutExpired:
        print("Error: openssl timed out after 15 seconds.", file=sys.stderr)
        sys.exit(1)


def cmd_info(args: argparse.Namespace) -> None:
    """Show info about a CMS signature file."""
    sig_path = Path(args.signature)
    if not sig_path.exists():
        print(f"Error: {sig_path} not found", file=sys.stderr)
        sys.exit(1)

    try:
        file_size = sig_path.stat().st_size
    except OSError as e:
        print(f"Error: cannot access {sig_path}: {e}", file=sys.stderr)
        sys.exit(1)
    print(f"Signature: {sig_path.name} ({file_size} bytes)")

    try:
        sig_bytes = sig_path.read_bytes()
    except OSError as e:
        print(f"Error reading {sig_path}: {e}", file=sys.stderr)
        sys.exit(1)
    try:
        content_info = asn1_cms.ContentInfo.load(sig_bytes)
        signed_data = content_info["content"]
        certs = signed_data["certificates"]
    except (ValueError, TypeError, KeyError, OSError) as e:
        print(f"  Error parsing signature: {e}", file=sys.stderr)
        return

    if not certs:
        print("  No certificates found in signature.")
        return

    cert_count = len(certs)
    print(f"\nCertificates ({cert_count}):")
    for i in range(cert_count):
        cert = certs[i].chosen
        if cert_count > 1:
            print(f"\n  [{i + 1}]")
        print(f"  Subject: {cert.subject.human_friendly}")
        print(f"  Issuer:  {cert.issuer.human_friendly}")
        print(f"  Serial:  {cert.serial_number}")
        print(
            f"  Valid:   {cert['tbs_certificate']['validity']['not_before'].native}"
            f" - {cert['tbs_certificate']['validity']['not_after'].native}"
        )

File: typescript/src/cli/verify.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * Signature verification and inspection commands.
 *
 * cmdVerify uses openssl CLI (graceful degradation if not installed).
 * cmdInfo uses pkijs (cross-platform, no external tools).
 */

import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, statSync } from "node:fs";

import { formatExpirySummary } from "../core/cert-expiry.js";
import { extractCertInfoFromCms } from "../core/cert-info.js";
import { getErrorMessage, isNodeError } from "../errors.js";

// -- Detached verify (openssl) ------------------------------------------------

export interface VerifyCommandArgs {
  pdf: string;
  signature?: string | null;
}

export function cmdVerify(args: VerifyCommandArgs): void {
  const pdfPath = args.pdf;
  const sigPath = args.signature ?? `${pdfPath}.p7s`;

  if (!existsSync(pdfPath)) {
    process.stderr.write(`Error: ${pdfPath} not found\n`);
    process.exit(1);
  }
  if (!existsSync(sigPath)) {
    process.stderr.write(`Error: ${sigPath} not found\n`);
    process.exit(1);
  }

  console.log(`Verifying ${pdfPath} against ${sigPath}...`);

  const opensslBin = "openssl";
  const opensslArgs = [
    "cms",
    "-verify",
    "-inform",
    "DER",
    "-in",
    sigPath,
    "-content",
    pdfPath,
    "-binary",
    "-purpose",
    "any",
  ];

  console.log("  Using system trust store for chain verification");

  try {
    const result = execFileSync(opensslBin, opensslArgs, {
      timeout: 15_000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    const stdout = result.toString("utf-8").trim();
    console.log("  VALID: Signature verification succeeded.");
    if (stdout.includes("Verification successful")) {
      console.log(`  ${stdout}`);
    }
  } catch (e: unknown) {
    if (isNodeError(e, "ENOENT")) {
      process.stderr.write("Error: openssl not found. Install OpenSSL to verify signatures.\n");
      process.exit(1);
    }
    // execFileSync attaches stderr buffer on non-zero exit
    if (e instanceof Error && "stderr" in e) {
      const { stderr } = e;
      if (Buffer.isBuffer(stderr)) {
        const stderrText = stderr.toString("utf-8").trim();
        console.log(`  INVALID: ${stderrText}`);
        process.exit(1);
      }
    }
    throw e;
  }
}

// -- CMS info (pkijs) ---------------------------------------------------------

export interface InfoCommandArgs {
  signature: string;
}

export async function cmdInfo(args: InfoCommandArgs): Promise<void> {
  const sigPath = args.signature;
  if (!existsSync(sigPath)) {
    process.stderr.write(`Error: ${sigPath} not found\n`);
    process.exit(1);
  }

  let fileSize: number;
  try {
    fileSize = statSync(sigPath).size;
  } catch (e) {
    process.stderr.write(`Error: cannot access ${sigPath}: ${getErrorMessage(e)}\n`);
    process.exit(1);
  }
  console.log(`Signature: ${sigPath} (${fileSize} bytes)`);

  let sigBytes: Uint8Array;
  try {
    sigBytes = new Uint8Array(readFileSync(sigPath));
  } catch (e) {
    process.stderr.write(`Error reading ${sigPath}: ${getErrorMessage(e)}\n`);
    process.exit(1);
  }

  try {
    const info = extractCertInfoFromCms(sigBytes);
    console.log("\nCertificate info:");
    if (info.name) console.log(`  Subject: ${info.name}`);
    if (info.organization) console.log(`  Organization: ${info.organization}`);
    if (info.email) console.log(`  Email: ${info.email}`);
    if (info.dn) console.log(`  DN: ${info.dn}`);
    if (info.notAfter) {
      console.log(`  Status: ${formatExpirySummary(info.notAfter)}`);
    }
  } catch (e) {
    process.stderr.write(`  Error parsing signature: ${getErrorMessage(e)}\n`);
  }
}
